### PR TITLE
BLD: add back support for building against numpy 1.x

### LINKED ISSUE #2644: 3.14 enforces numpy >= 2, despite comment in setup.py
The `pyproject.toml` clearly requires `numpy >=2` to build, [however setup.py notes](https://github.com/h5py/h5py/blob/3.14.0/setup.py#L31):
```
 # here, and if you can build an older NumPy on a newer Python, h5py probably
    # works (assuming you build it from source too).
    # NumPy 1.19.3 is the first with wheels for Python 3.9, our minimum Python.
```

This worked pre-3.14, however, [3.14 looks to enforce](https://github.com/h5py/h5py/blob/3.14.0/setup_build.py#L130) the build with `numpy >= 2`
```
assert numpy.__version__ >= '2.0'  # See build dependencies in pyproject.toml
```

I wanted to confirm that builds of >=3.14 really do require `numpy >= 2`

### PR DESCRIPTION
**Pull request by neutrinoceros**
Close #2644
At the time of opening, there's a missing piece: I haven't implemented a warning for `h5py._npystrings` being missing in this configuration yet.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
I'll need to iron out a couple defects too:
- mindeps is broken on every normal build
- pre-deps cannot possibly work with this build, I think

**Comment by neutrinoceros** [comment]
I think I managed to resolve immediate problems. Now I should only need to add and tests for the warning.
I'll be back to this soon.

**Comment by crusaderky** [comment]
The intersection of cython==0.29.37 and numpy>=2 segfaults on npystrings (see https://github.com/h5py/h5py/pull/2647).
CI passes because it tests cython==0.29.37 and numpy<2.
Cython 3.0 fixes it.

I'd recommend adding a test for minimum version of cython + numpy==2.0.0 in addition to numpy 1.x.

File: setup_build.py [hunk 1]
```diff
@@ -117,22 +117,27 @@ def _make_extensions(cls, config):
...
             settings['runtime_library_dirs'] = settings['library_dirs']
 
-        return [cls._make_extension(m, settings) for m in MODULES]
+        # 'NPY_1_21_API_VERSION' == 0x0000000e
+        settings['define_macros'].append(('NPY_TARGET_VERSION', 'NPY_1_21_API_VERSION'))
+        extensions = [cls._make_extension(m, settings) for m in MODULES]
+
+        if int(numpy.__version__.split('.')[0]) >= 2:
+            # Enable NumPy 2.0 C API for modules that require it.
+            # NUMPY2_MODULES will not be importable when NumPy 1.x is installed.
```
Comment (crusaderky) [inline setup_build.py:126]:
```suggestion
            # MODULES_NUMPY2_ONLY will not be importable when NumPy 1.x is installed.
```

File: setup_build.py [hunk 2]
```diff
@@ -117,22 +117,27 @@ def _make_extensions(cls, config):
...
+            # 'NPY_2_0_API_VERSION' == 0x00000012
+            settings['define_macros'].append(('NPY_TARGET_VERSION', 'NPY_2_0_API_VERSION'))
+            extensions.extend(cls._make_extension(m, settings) for m in MODULES_NUMPY2_ONLY)
```
Comment (crusaderky) [inline setup_build.py]:
```suggestion
        extensions = [cls._make_extension(m, settings) for m in MODULES]

        if int(numpy.__version__.split('.')[0]) >= 2:
            # Enable NumPy 2.0 C API for modules that require it.
            # MODULES_NUMPY2_ONLY will not be importable when NumPy 1.x is installed.
            # 'NPY_2_0_API_VERSION' == 0x00000012
            settings['define_macros'].append(('NPY_TARGET_VERSION', 'NPY_2_0_API_VERSION'))
            extensions.extend(cls._make_extension(m, settings) for m in MODULES_NUMPY2_ONLY)
        else:
            # 'NPY_1_21_API_VERSION' == 0x0000000e
            settings['define_macros'].append(('NPY_TARGET_VERSION', 'NPY_1_21_API_VERSION'))
```

Comment (neutrinoceros) [reply setup_build.py]: Good point, better build all extensions with a consistent `NPY_TARGET_VERSION`. Not sure your suggestion applies as is though so I'll write my own version of it. Thanks !

**Comment by neutrinoceros** [comment]
> I'd recommend adding a test for minimum version of cython + numpy==2.0.0 in addition to numpy 1.x.

great suggestion, thanks a lot. I'll let CI run its course without it first since I just did a "risky" rebase, but I'll include this in the next iteration !

File: setup_build.py (hunk 2 above)
Comment (neutrinoceros) [reply setup_build.py]: Actually this doesn't seem to work; the resulting wheels are not compatible with numpy 1.21, so I reverted this change.

https://github.com/h5py/h5py/actions/runs/16965494246/job/48088254143?pr=2646

**Comment by neutrinoceros** [comment]
So I started working on implementing a warning as suggested in #2644, until I realized that importing h5py, built with numpy 1.x, and with numpy 2.x present at runtime, would already raise an exception:
```python-traceback
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/clm/dev/h5py/h5py/__init__.py", line 45, in <module>
    from ._conv import register_converters as _register_converters, \
  File "h5py/_conv.pyx", line 1, in init h5py._conv
    # cython: profile=False
  File "h5py/h5r.pyx", line 1, in init h5py.h5r
    # cython: language_level=3
  File "h5py/h5p.pyx", line 1, in init h5py.h5p
    # cython: language_level=3
ValueError: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject
```

with this in mind, I don't think adding a warning is necessary or helpful; this setup is clearly broken, we shouldn't need to tell users that it is.
It's also clear from the traceback where the problem is coming from.

File: ci/cibw_test_command.sh [hunk 3]
```diff
@@ -12,15 +12,20 @@ WHEEL_PATH=$2
...
+if [[ "$ONLY_MINDEPS_TESTS" == "1" ]] ; then
+    ENVLIST="py$PYVER-test-mindeps"
+else
+    ENVLIST="py$PYVER-test-deps,py$PYVER-test-deps-pre"
+    if ! ( [[ "$AUDITWHEEL_PLAT" == musllinux_* ]] && [[ "$PYVER" == "39" || "$PYVER" == "310" || "$PYVER" == "311" ]] ); then
+        # skip mindeps on musllinux + python 3.9 to 3.11 because oldest supported numpy versions
+        # are higher for this target (1.25.0) and there's no way (that I found) to specify it
+        # directly in package metadata.
+        ENVLIST="py$PYVER-test-mindeps,$ENVLIST"
+    fi
```
Comment (neutrinoceros) [inline ci/cibw_test_command.sh]: this can be simplified a lot with #2631

**Comment by neutrinoceros** [comment]
@takluyver what do you think ?

**Comment by neutrinoceros** [comment]
rebased to resolve a merge conflict

File: .github/workflows/build_wheels.yml [hunk 4]
```diff
@@ -103,9 +103,55 @@ jobs:
...
+      # Get HDF5 outside of cibuildwheel so that we can set env vars via GITHUB_ENV
+      # inside the HDF5-getting scripts and have them be in effect for the
+      # cibuildwheel step.
+
+      # Cache HDF5
+      - name: Cache built HDF5
+        uses: actions/cache@0400d5f644dc74513175e3cd8d07132dd4860809 # v4.2.4
+        with:
+          key: ${{ matrix.os }}-${{ matrix.arch }}-6
+          path: cache/hdf5
```
Comment (takluyver) [inline .github/workflows/build_wheels.yml]: I think this cache step is doing nothing here - in the main `build_wheels` job it caches HDF5 built on Windows & Mac, but this job only runs on Linux, and it gets HDF5 from the container image we point cibuildwheel at.

**Comment by takluyver** [comment]
> It's also clear from the traceback where the problem is coming from.

I don't think the `numpy.dtype size changed` error message is clear at all, except that people may have seen it before and remember what sort of thing caused it. But I don't expect this scenario to come up often, since we'll continue building packages with Numpy 2.x, so I agree the error is good enough and we don't need to add a warning.

File: .github/workflows/build_wheels.yml (hunk 4 above)
Comment (neutrinoceros) [reply .github/workflows/build_wheels.yml]: oh, nice catch ! it looks like I just copy-pasted this bit from above, where it's not conditional on the os (but could be). I'll delete it here and open a separate PR to skip this step where we don't need it !

**Comment by neutrinoceros** [comment]
> I don't think the numpy.dtype size changed error message is clear at all

you're right, I wasn't realising I made an assumption in my interpretation of it.

File: .github/workflows/build_wheels.yml (hunk 4 above)
Comment (neutrinoceros) [reply .github/workflows/build_wheels.yml]: (cleaned up here)

Comment (neutrinoceros) [reply .github/workflows/build_wheels.yml]: and other PR at https://github.com/h5py/h5py/pull/2695

File: news/bld-new-mindeps.rst [hunk 5]
```diff
@@ -0,0 +1,6 @@
+Building h5py
+-------------
+
+* The minimum versions build-time Python requirements were updated to
+  ``Cython==3.0.0`` (up from ``0.29.1``), and ``numpy==1.25.0`` (down from
+  ``2.0.0``).
```
Comment (takluyver) [inline news/bld-new-mindeps.rst]: Let's mention here that we still recommend building with Numpy 2.x in general, this is just to allow ecosystems where you build with the same version you want to use to still use Numpy 1.x.

**Comment by takluyver** [comment]
Other than that, I think this looks good.

File: news/bld-new-mindeps.rst (hunk 5 above)
Comment (neutrinoceros) [reply news/bld-new-mindeps.rst]: good idea. How's my attempt at it ?

File: news/bld-new-mindeps.rst [hunk 6]
```diff
@@ -0,0 +1,9 @@
+Building h5py
+-------------
+
+* The minimum versions build-time Python requirements were updated to
+  ``Cython==3.0.0`` (up from ``0.29.1``), and ``numpy==1.25.0`` (down from
+  ``2.0.0``). We still recommend building with numpy 2 or newer whenever
+  possible, this is done to improve support for external package ecosystems,
+  not controlled by h5py's maintainers, where numpy 2 might not be available
+  yet.
```
Comment (takluyver) [inline news/bld-new-mindeps.rst]: I think it's not so much about numpy 2 not being *available*, more:

```suggestion
  where the system uses the same version of packages for building and
  installation.
```

(You might want/need to install with Numpy 1.x, even if the ecosystem has 2.x available, because of some other dependency. The big conda environments we set up at work still get numpy 1.x, for instance, although with conda that doesn't mean you need to build with 1.x.)
